File: app/dashboard/static/js/triggers.js
/* Triggers — event-driven automations (list + firing log; authoring is agent-via-chat) */
(function() {
    const tbody = document.getElementById('trg-tbody');
    const logTbody = document.getElementById('trg-log-tbody');

    window.addEventListener('tab:triggers', loadAll);
    document.getElementById('trg-refresh-btn').addEventListener('click', loadAll);

    async function loadAll() {
        loadTriggers();
        loadFirings();
    }

    async function loadTriggers() {
        const data = await api('/api/triggers');
        if (!data) return;
        document.getElementById('trg-empty').style.display = data.length ? 'none' : '';
        tbody.innerHTML = data.map(t => {
            const statusCls = t.enabled ? 'green' : 'red';
            const last = t.last_fired ? new Date(t.last_fired * 1000).toLocaleString() : 'Never';
            const srcBadge = t.source_type === 'webhook' ? '&#9889; webhook' : '&#8634; poll';
            return `<tr>
                <td><span class="status-dot" style="background:var(--${statusCls});display:inline-block"></span>
                    <span style="font-weight:500">${esc(t.name)}</span>
                    <div style="font-size:11px;color:var(--text-dim)">${esc(t.description)}</div></td>
                <td><code style="font-size:11px">${srcBadge}</code></td>
                <td style="max-width:260px;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;font-size:12px">${esc(t.action)}</td>
                <td style="font-size:12px">${t.sinks.map(esc).join(', ')}</td>
                <td style="max-width:180px;overflow:hidden;text-overflow:ellipsis;white-space:nowrap"><code style="font-size:11px">${esc(t.filter)}</code></td>
                <td style="font-size:11px;color:var(--text-dim)">${last}</td>
                <td>
                    <button class="action-btn" onclick="trgToggle('${t.name}')">${t.enabled ? 'Disable' : 'Enable'}</button>
                    <button class="action-btn" onclick="trgTest('${t.name}')">Test</button>
                    <button class="action-btn delete" onclick="trgDelete('${t.name}')">Del</button>
                </td>
            </tr>`;
        }).join('');
    }

    async function loadFirings() {
        const rows = await api('/api/triggers/firings?limit=50');
        if (!rows) return;
        document.getElementById('trg-log-empty').style.display = rows.length ? 'none' : '';
        const colors = { fired: 'green', error: 'red', filtered: 'text-dim', debounced: 'yellow' };
        logTbody.innerHTML = rows.map(r => {
            const c = colors[r.status] || 'text-dim';
            const result = (r.result || '').slice(0, 160);
            const dur = r.duration_ms != null ? (r.duration_ms / 1000).toFixed(1) + 's' : '';
            return `<tr>
                <td style="font-size:11px;color:var(--text-dim);white-space:nowrap">${new Date(r.timestamp * 1000).toLocaleString()}</td>
                <td style="font-weight:500">${esc(r.trigger_name)}</td>
                <td style="font-size:12px">${esc(r.source)}</td>
                <td><span style="color:var(--${c})">${esc(r.status)}</span></td>
                <td style="max-width:380px;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;font-size:12px" title="${esc(r.result || '')}">${esc(result)}</td>
                <td style="font-size:11px;color:var(--text-dim)">${dur}</td>
            </tr>`;
        }).join('');
    }

    function esc(s) {
        const d = document.createElement('div');
        d.textContent = s || '';
        return d.innerHTML;
    }

    window.trgToggle = async function(name) {
        await api(`/api/triggers/${name}/toggle`, { method: 'POST' });
        loadTriggers();
    };

    window.trgTest = async function(name) {
        const raw = prompt('Test payload (JSON):', '{"event": "test"}');
        if (raw === null) return;
        let payload = {};
        try { payload = JSON.parse(raw || '{}'); } catch { alert('Invalid JSON'); return; }
        const res = await api(`/api/triggers/${name}/test`, {
            method: 'POST',
            body: JSON.stringify(payload),
        });
        if (res) alert(`Status: ${res.status} — check Recent firings in a few seconds.`);
        setTimeout(loadFirings, 1500);
    };

    window.trgDelete = async function(name) {
        if (!confirm(`Delete trigger '${name}'?`)) return;
        await api(`/api/triggers/${name}`, { method: 'DELETE' });
        loadTriggers();
    };
})();

File: app/dashboard/routes/triggers.py
"""Trigger management API — list/inspect/toggle/delete triggers and read the
firing history. Session-authenticated (the dashboard UI's Triggers tab).
Authoring happens via the triggers MCP server (the agent), not here, in v1."""

from quart import Blueprint, jsonify, request, current_app

import db
import trigger_store
from dashboard.auth import require_auth

triggers_bp = Blueprint("triggers", __name__)


def _summary(t: dict, last_fired: dict[str, float]) -> dict:
    act = t.get("action") or {}
    action_label = (f"workflow: {act.get('workflow')}" if act.get("type") == "workflow"
                    else f"prompt: {(act.get('prompt') or '')[:60]}")
    return {
        "name": t.get("name"),
        "description": t.get("description", ""),
        "source_type": (t.get("source") or {}).get("type"),
        "action": action_label,
        "sinks": t.get("sinks") or ["telegram"],
        "enabled": t.get("enabled", True),
        "debounce_seconds": t.get("debounce_seconds", 0),
        "filter": t.get("filter") or "",
        "last_fired": last_fired.get(t.get("name")),
    }


@triggers_bp.route("/api/triggers")
@require_auth
async def list_triggers():
    last = db.get_last_fired_map()
    return jsonify([_summary(t, last) for t in trigger_store.load_triggers()])


@triggers_bp.route("/api/triggers/<name>")
@require_auth
async def get_trigger(name):
    t = trigger_store.get_trigger(name)
    if not t:
        return jsonify({"error": "not found"}), 404
    out = dict(t)
    if (t.get("source") or {}).get("type") == "webhook":
        out["webhook_url"] = trigger_store.webhook_url(name)
    return jsonify(out)


@triggers_bp.route("/api/triggers/<name>/toggle", methods=["POST"])
@require_auth
async def toggle_trigger(name):
    t = trigger_store.get_trigger(name)
    if not t:
        return jsonify({"error": "not found"}), 404
    enabled = not t.get("enabled", True)
    trigger_store.set_enabled(name, enabled)
    return jsonify({"name": name, "enabled": enabled})


@triggers_bp.route("/api/triggers/<name>", methods=["DELETE"])
@require_auth
async def delete_trigger(name):
    if not trigger_store.delete_trigger(name):
        return jsonify({"error": "not found"}), 404
    return jsonify({"deleted": name})


@triggers_bp.route("/api/triggers/firings")
@require_auth
async def all_firings():
    limit = min(int(request.args.get("limit", 50)), 500)
    return jsonify(db.get_trigger_firings(limit=limit))


@triggers_bp.route("/api/triggers/<name>/firings")
@require_auth
async def trigger_firings(name):
    limit = min(int(request.args.get("limit", 50)), 500)
    return jsonify(db.get_trigger_firings(trigger_name=name, limit=limit))


@triggers_bp.route("/api/triggers/<name>/test", methods=["POST"])
@require_auth
async def test_trigger(name):
    t = trigger_store.get_trigger(name)
    if not t:
        return jsonify({"error": "not found"}), 404
    payload = await request.get_json(silent=True) or {}
    status = current_app.trigger_engine.accept(t, payload, source="test")
    return jsonify({"status": status})
